app/classes/query/modules/lookup_names.rb

```diff
@@ -97,59 +97,91 @@ def add_other_spellings(min_names)
         ids = min_names.map { |min_name| min_name[1] || min_name[0] }
         return [] if ids.empty?
 
-        Name.connection.select_rows(%(
-          SELECT #{minimal_name_columns} FROM names
-          WHERE COALESCE(correct_spelling_id, id) IN (#{clean_id_set(ids)})
-        ))
+        Name.
+          where((Name[:correct_spelling_id].coalesce(Name[:id])).
+                in(clean_id_set(ids).split(",").map(&:to_i))).
+          pluck(*minimal_name_columns)
       end
 
       def add_synonyms(min_names)
         ids = min_names.map { |min_name| min_name[2] }.reject(&:nil?)
         return min_names if ids.empty?
 
         min_names.reject { |min_name| min_name[2] } +
-          Name.connection.select_rows(%(
-            SELECT #{minimal_name_columns} FROM names
-            WHERE synonym_id IN (#{clean_id_set(ids)})
-          ))
+          Name.where(synonym_id: clean_id_set(ids).split(",")).
+          pluck(*minimal_name_columns)
       end
 
       def add_subtaxa(min_names)
         higher_names = genera_and_up(min_names)
         lower_names = genera_and_down(min_names)
-        unless higher_names.empty?
-          regex = escape(": _(#{higher_names.join("|")})_")
-          min_names += Name.connection.select_rows(%(
-            SELECT #{minimal_name_columns} FROM names
-            WHERE classification REGEXP #{regex}
-          ))
+
+        if higher_names.empty?
+          include_lower_taxa(
+            names: min_names,
+            lower_names_regex: /^(#{lower_names.join("|")}) /
+          )
+        else
+          include_higher_and_lower_taxa(
+            names: min_names,
+            lower_names_regex: /^(#{lower_names.join("|")}) /,
+            higher_names_matcher: /: _(#{higher_names.join("|")})_/
+          )
         end
-        regex = escape("^(#{lower_names.join("|")}) ")
-        min_names += Name.connection.select_rows(%(
-          SELECT #{minimal_name_columns} FROM names
-          WHERE text_name REGEXP #{regex}
-        ))
-        min_names.uniq
+      end
+
+      def include_lower_taxa(names:, lower_names_regex:)
+        # The names_themselves
+        Name.where(id: names.map(&:first)).
+          # + lower names
+          or(Name.where(Name[:text_name] =~ lower_names_regex)).
+          distinct.
+          pluck(*minimal_name_columns)
+      end
+
+      def include_higher_and_lower_taxa(names:, lower_names_regex:,
+                                        higher_names_matcher:)
+        # The names_themselves
+        Name.where(id: names.map(&:first)).
+          # + higher names
+          or(Name.where(Name[:classification] =~ higher_names_matcher)).
+          # + lower names
+          or(Name.where(Name[:text_name] =~ lower_names_regex)).
+          distinct.
+          pluck(*minimal_name_columns)
```

Comment:
This works, and as you say, it reduces the number of queries from two to one in some cases.  I'm unclear on whether moving the processing of "distinct" from the webserver to the db is a good idea, but I doubt this is a huge performance sink.  There is a certain elegance to your approach that I do like.

Did you consider approaching it as building a query one step at a time instead of creating the entire query all in one go?  It may help DRY this up a bit.  I'm thinking something along these lines:

`def add_subtaxa(min_names)`
`  higher_names = genera_and_up(min_names)`
`  lower_names = genera_and_down(min_names)`
`  query = Name.where(id: min_names.map(&:first))`
`  query = add_lower_names(query, lower_names)`
`  query = add_higher_names(query, higher_names)`
`  query.distinct.pluck(*minimal_name_columns)`
`end`

`def add_lower_names(query, names)`
`  query.or(Name[:text_name] =~ /^(#{names.join("|")}) /)`
`end`

`def add_higher_names(query, names)`
`  query.or(Name[:classification] =~ /: _(#{names.join("|")})_/)`
`end`